pkg/wallet/storage.go

```diff
@@ -0,0 +1,139 @@
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/bluele/gcache"
+
+	"github.com/hyperledger/aries-framework-go/component/storage/edv"
+	"github.com/hyperledger/aries-framework-go/pkg/crypto"
+	"github.com/hyperledger/aries-framework-go/pkg/crypto/tinkcrypto"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/packer"
+	"github.com/hyperledger/aries-framework-go/pkg/doc/jose"
+	"github.com/hyperledger/aries-framework-go/pkg/kms"
+	"github.com/hyperledger/aries-framework-go/spi/storage"
+)
+
+// storageProvider all wallet content store storage provider operations.
+// if profile has EDV settings then call to OpenStore will return EDV store instance by creating new EDV provider.
+// Otherwise default provider will be used to open store.
+// (Refer #2745 for more details)
+type storageProvider struct {
+	profile         *profile
+	defaultProvider storage.Provider
+}
+
+func newWalletStorageProvider(profile *profile, provider storage.Provider) *storageProvider {
+	return &storageProvider{profile: profile, defaultProvider: provider}
+}
+
+// OpenStore opens and returns store and sets store config to provider.
+// if wallet profile has EDV settings then auth provided will be used to initialize edv storage provider.
+func (s *storageProvider) OpenStore(auth string, config storage.StoreConfiguration) (storage.Store, error) {
+	var provider storage.Provider
+	var err error
+
+	if s.profile.EDVConf != nil {
+		provider, err = createEDVStorageProvider(auth, s.profile.EDVConf)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		provider = s.defaultProvider
+	}
+
+	store, err := provider.OpenStore(s.profile.ID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open store : %w", err)
+	}
+
+	err = provider.SetStoreConfig(s.profile.ID, config)
+	if err != nil {
+		e := store.Close()
+		if e != nil {
+			logger.Warnf("failed to close store: %s", e)
+		}
+
+		return nil, fmt.Errorf("failed to set store config: %w", err)
+	}
+
+	return store, nil
+}
+
+func createEDVStorageProvider(auth string, conf *edvConf) (storage.Provider, error) {
+	// get key manager
+	keyMgr, err := keyManager().getKeyManger(auth)
+	if err != nil {
+		if errors.Is(err, gcache.KeyNotFoundError) {
+			return nil, ErrWalletLocked
+		}
+
+		return nil, err
+	}
+
+	// get crypto
+	cryptoImpl, err := tinkcrypto.New()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create crypto: %w", err)
+	}
+
+	// get jwe encrypter
+	jweEncrypter, err := getJWSEncrypter(conf.EncryptionKeyID, keyMgr, cryptoImpl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create JWE encrypter: %w", err)
+	}
+
+	// get jwe decrypter
+	jweDecrypter := jose.NewJWEDecrypt(nil, cryptoImpl, keyMgr)
+
+	// get MAC crypto
+	macCrypto, err := getMacCrypto(conf.MACKeyID, keyMgr, cryptoImpl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create mac crypto: %w", err)
+	}
+
+	// create EDV provider
+	// TODO add zcap header support
+	return edv.NewRESTProvider(conf.ServerURL, conf.VaultID,
```

Comment:
Those options `edv.WithDeterministicDocumentIDs()),edv.WithFullDocumentsReturnedFromQueries(), edv.WithBatchEndpointExtension()` should be configurable since they're not standard - especially the last two options, which only work with our TrustBloc EDV server implementation.

(If it's too much work right now, I would suggest creating a followup issue)

Reply:
created a followup